algebra-core/src/fields/models/fp4.rs

```diff
@@ -0,0 +1,427 @@
+    }
+}
+
+#[derive(Derivative)]
+#[derivative(
+    Default(bound = "P: Fp4Parameters"),
+    Hash(bound = "P: Fp4Parameters"),
+    Clone(bound = "P: Fp4Parameters"),
+    Copy(bound = "P: Fp4Parameters"),
+    Debug(bound = "P: Fp4Parameters"),
+    PartialEq(bound = "P: Fp4Parameters"),
+    Eq(bound = "P: Fp4Parameters")
+)]
+pub struct Fp4<P: Fp4Parameters> {
+    pub c0:          Fp2<P::Fp2Params>,
+    pub c1:          Fp2<P::Fp2Params>,
+    #[derivative(Debug = "ignore")]
+    #[doc(hidden)]
+    pub _parameters: PhantomData<P>,
+}
+
+impl<P: Fp4Parameters> Fp4<P> {
+    pub fn new(c0: Fp2<P::Fp2Params>, c1: Fp2<P::Fp2Params>) -> Self {
+        Fp4 {
+            c0,
+            c1,
+            _parameters: PhantomData,
+        }
+    }
+
+    pub fn unitary_inverse(&self) -> Self {
+        Self::new(self.c0, -self.c1)
+    }
+
+    pub fn cyclotomic_exp<B: BigInteger>(&self, exponent: &B) -> Self {
+        let mut res = Self::one();
+        let self_inverse = self.unitary_inverse();
+
+        let mut found_nonzero = false;
+        let naf = exponent.find_wnaf();
+
+        for &value in naf.iter().rev() {
+            if found_nonzero {
+                res = res.square();
+            }
+
+            if value != 0 {
+                found_nonzero = true;
+
+                if value > 0 {
+                    res *= self;
+                } else {
+                    res *= &self_inverse;
+                }
+            }
+        }
+
+        res
+    }
+
+    pub fn mul_by_fp(&mut self, element: &<P::Fp2Params as Fp2Parameters>::Fp) {
+        self.c0.mul_assign_by_fp(element);
+        self.c1.mul_assign_by_fp(element);
+    }
+
+    pub fn mul_by_fp2(&mut self, element: &Fp2<P::Fp2Params>) {
+        self.c0 *= element;
+        self.c1 *= element;
+    }
+}
+
+impl<P: Fp4Parameters> Zero for Fp4<P> {
+    fn zero() -> Self {
+        Fp4::new(<Fp2<P::Fp2Params>>::zero(), <Fp2<P::Fp2Params>>::zero())
+    }
+
+    fn is_zero(&self) -> bool {
+        self.c0.is_zero() && self.c1.is_zero()
+    }
+}
+
+impl<P: Fp4Parameters> One for Fp4<P> {
+    fn one() -> Self {
+        Fp4::new(<Fp2<P::Fp2Params>>::one(), <Fp2<P::Fp2Params>>::zero())
+    }
+
+    fn is_one(&self) -> bool {
+        self.c0.is_one() && self.c1.is_zero()
+    }
+}
+
+impl<P: Fp4Parameters> Field for Fp4<P> {
+    #[inline]
+    fn characteristic<'a>() -> &'a [u64] {
+        <P::Fp2Params as Fp2Parameters>::Fp::characteristic()
+    }
+
+    fn double(&self) -> Self {
+        let mut result = self.clone();
+        result.double_in_place();
+        result
+    }
+
+    fn double_in_place(&mut self) -> &mut Self {
+        self.c0.double_in_place();
+        self.c1.double_in_place();
+        self
+    }
+
+    fn square(&self) -> Self {
+        let mut result = *self;
+        result.square_in_place();
+        result
+    }
+
+    fn square_in_place(&mut self) -> &mut Self {
+        // v0 = c0 - c1
```

Comment:
Can you add a link to a reference/citation?

Reply:
It is taken from the corresponding Fp2 computation.
While it computes the same formula as in "Devegili OhEig Scott Dahab – Multiplication and Squaring on AbstractPairing-Friendly Fields", it differs slightly in its representation.
For more clarity, I replaced the code in Fp4 by the exact formula as in the reference above (will be in my next commit).

In particular, the code here computes
```
c0 = (a0 - a1)(a0 - βa1) + v0 + βv0
c1 = 2v0
```
whereas the reference uses
```
c0 =(a0 + a1)(a0 + βa1) − v0 − βv0
c1 = 2v0,
```
with `v0 = a0 * a1`.
Both are correct though.